### PR TITLE
Add initial User Schema and GET, POST, PUT, DELETE Endpoints

### PR DESCRIPTION
**Pull request by amit-sawhney**
## Summary

Creates a User Schema and the basic GET, POST, PUT, and DELETE requests related to a User
Refs #17 

## Changes

- Add GET All users Endpoint `/api/users/
- Add GET Specific User By ID Endpoint `/api/users/:userId`
- Add POST a new user `/api/users/`
- Add PUT to a specific user  `/api/users/:userId`
- Add DELETE a specific user `/api/users/:userId`
- Add testing for endpoints

## Requests / Responses

**Request**
GET `/api/users` Returns a list of users

**Response**
HTTP/1.1 200 OK

```
{
    "message": "Successfully retrieved all users.",
    "success": true,
    "result": [
        {
            "firstName": "Alice",
            "middleName": null,
            "lastName": "Fang",
            "preferredName": "AF",
            "email": "af@gmail.com",
            "phone": null,
            "oauthID": "2",
            "genders": [],
            "pronouns": null,
            "masthead": false,
            "portfolio": null,
            "linkedIn": null,
            "twitter": null,
            "claimedPitches": [],
            "submittedPitches": [],
            "currentTeams": [],
            "role": "CONTRIBUTOR",
            "races": [],
            "interests": [
                "CANNABIS"
            ],
            "_id": "6031a866c70ec705736a79e5",
            "gender": null,
            "dateJoined": "2021-02-21T00:25:10.279Z",
            "__v": 0
        },
        {
            "firstName": "Amit",
            "middleName": null,
            "lastName": "Sawhney",
            "preferredName": "AF",
            "email": "af@gmail.com",
            "phone": null,
            "oauthID": "3",
            "genders": [],
            "pronouns": null,
            "masthead": false,
            "portfolio": null,
            "linkedIn": null,
            "twitter": null,
            "claimedPitches": [],
            "submittedPitches": [],
            "currentTeams": [],
            "role": "CONTRIBUTOR",
            "races": [],
            "interests": [
                "CANNABIS"
            ],
            "_id": "6031a92ac70ec705736a79e8",
            "gender": null,
            "dateJoined": "2021-02-21T00:28:26.769Z",
            "__v": 0
        },
        {
            "firstName": "Alice",
            "middleName": null,
            "lastName": "Fang",
            "preferredName": "AF",
            "email": "af@gmail.com",
            "phone": null,
            "oauthID": "4",
            "genders": [],
            "pronouns": null,
            "masthead": false,
            "portfolio": null,
            "linkedIn": null,
            "twitter": null,
            "claimedPitches": [],
            "submittedPitches": [],
            "currentTeams": [],
            "role": "TBD",
            "races": [],
            "interests": [
                "NONE"
            ],
            "_id": "6031aac903fb1305fad158f3",
            "gender": null,
            "dateJoined": "2021-02-21T00:35:21.610Z",
            "__v": 0
        },
        {
            "firstName": "Development",
            "middleName": null,
            "lastName": "User",
            "preferredName": null,
            "email": null,
            "phone": null,
            "oauthID": null,
            "genders": [],
            "pronouns": [],
            "masthead": false,
            "portfolio": null,
            "linkedIn": null,
            "twitter": null,
            "claimedPitches": [],
            "submittedPitches": [],
            "currentTeams": [],
            "role": "TBD",
            "races": [],
            "interests": [],
            "_id": "60332ef8dd742493ce5562b3",
            "dateJoined": "2021-02-24T18:00:35.296Z"
        }
    ]
}
```

**Request**
GET `/api/users/:userId` Returns  a specific user

**Response**
HTTP/1.1 200 OK

```
{
    "success": true,
    "result": {
        "firstName": "Alice",
        "middleName": null,
        "lastName": "Fang",
        "preferredName": "AF",
        "email": "af@gmail.com",
        "phone": null,
        "oauthID": "2",
        "genders": [],
        "pronouns": null,
        "masthead": false,
        "portfolio": null,
        "linkedIn": null,
        "twitter": null,
        "claimedPitches": [],
        "submittedPitches": [],
        "currentTeams": [],
        "role": "CONTRIBUTOR",
        "races": [],
        "interests": [
            "CANNABIS"
        ],
        "_id": "6031a866c70ec705736a79e5",
        "gender": null,
        "dateJoined": "2021-02-21T00:25:10.279Z",
        "__v": 0
    },
    "message": "Successfully retrieved user"
}
```

**Request**
Post `/api/users/new` Creates a new User
`
name: "PR Test User"
`

**Response**
HTTP/1.1 200 OK
```
{
    "message": "Successfully created new user",
    "success": true,
    "result": {
        "firstName": "PR",
        "middleName": null,
        "lastName": "Test",
        "preferredName": null,
        "email": null,
        "phone": null,
        "oauthID": null,
        "genders": [],
        "pronouns": [],
        "masthead": false,
        "portfolio": null,
        "linkedIn": null,
        "twitter": null,
        "claimedPitches": [],
        "submittedPitches": [],
        "currentTeams": [],
        "role": "TBD",
        "races": [],
        "interests": [],
        "_id": "60369484afa92444e4ab8aa3",
        "dateJoined": "2021-02-24T18:01:40.761Z",
        "__v": 0
    }
}
```

**Request**
PUT`/api/users/:userId` Updates  a specific user

**Response**
HTTP/1.1 200 OK

```
{
    "success": true,
    "message": "Successfully updated user",
    "result": {
        "firstName": "PR",
        "middleName": null,
        "lastName": "Change",
        "preferredName": null,
        "email": null,
        "phone": null,
        "oauthID": null,
        "genders": [],
        "pronouns": [],
        "masthead": false,
        "portfolio": null,
        "linkedIn": null,
        "twitter": null,
        "claimedPitches": [],
        "submittedPitches": [],
        "currentTeams": [],
        "role": "TBD",
        "races": [],
        "interests": [],
        "_id": "60369484afa92444e4ab8aa3",
        "dateJoined": "2021-02-24T18:01:40.761Z",
        "__v": 0
    }
}
```

**Request**
DELETE `/api/users/:userId` Returns  a specific user

**Response**
HTTP/1.1 200 OK

```
{
    "success": true,
    "message": "User successfully deleted",
    "result": {
        "firstName": "PR",
        "middleName": null,
        "lastName": "Change",
        "preferredName": null,
        "email": null,
        "phone": null,
        "oauthID": null,
        "genders": [],
        "pronouns": [],
        "masthead": false,
        "portfolio": null,
        "linkedIn": null,
        "twitter": null,
        "claimedPitches": [],
        "submittedPitches": [],
        "currentTeams": [],
        "role": "TBD",
        "races": [],
        "interests": [],
        "_id": "60369484afa92444e4ab8aa3",
        "dateJoined": "2021-02-24T18:01:40.761Z",
        "__v": 0
    }
}
```

**Request**
GET, PUT, DELETE `/api/users/{UNKNOWN_ID}` Returns a 404 error indicating user not found

**Response**
HTTP/1.1 404 Not Found

```
{
    "success": false,
    "message": "User not found with id"
}
```

**Request**
GET, PUT, DELETE `/api/users/{Bad_MongoID}` Returns a 400 error indicating the passed MongoID is invalid

**Response**
HTTP/1.1 400 Bad Request

```
{
    "success": false,
    "message": "Bad ID format"
}
```

### CODE REVIEW

**Review by sahithimuthyala — CHANGES_REQUESTED** [review]
Make adjustments to the user model based on modifications made to the [db design](https://docs.google.com/document/d/1auis10INg0Ep5fqLgdpqJ5fgPSzfRBLjufgaSbYNqNM/edit). @alicesf2 will be reviewing the comments + suggestions on the db design doc today (Tuesday) morning. Chat with her asynchronously and make sure schema is "finalized" before making additional changes.

File: api/src/api/users.ts [hunk 1]
```diff
@@ -1,20 +1,135 @@
...
+        result: user,
+        message: `Successfully retrieved user`,
+      });
+    }
+  }),
+);
+
+// Create a new user
+router.post(
+  '/new',
```
Comment (alicesf2) [inline api/src/api/users.ts:58]: I would avoid defining a separate endpoint for this - you can just post to /users since all your users live here

**Review by alicesf2 — CHANGES_REQUESTED** [review]
AWESOME WORK YOU GUYS 🤩  this is truly amazing. We made changes to the User schema in the DB doc as Sahi mentioned, so make sure to double-check the type of each field and add in the new fields as well (including the pink ones)

**Review by sahithimuthyala — CHANGES_REQUESTED** [review]
The following need to be added to the user model:
- race/ethnicity
- current team(s) -- ie. visuals
- masthead

File: api/src/models/user.ts [hunk 2]
```diff
@@ -1,11 +1,86 @@
...
+  submittedPitches: [{ type: Schema.Types.ObjectId, ref: 'Pitch' }],
+  currentTeams: [{ type: String, default: null }],
+
+  role: {
+    type: String,
+    enum: Object.values(rolesEnum),
+    default: rolesEnum.TBD,
+  },
+
+  interests: {
```
Comment (alicesf2) [inline api/src/models/user.ts:79]: Change this to an array of string enums

File: api/src/models/user.ts [hunk 3]
```diff
@@ -1,11 +1,86 @@
...
+  firstName: { type: String, default: null, required: true },
+  middleName: { type: String, default: null },
+  lastName: { type: String, default: null, required: true },
+  preferredName: { type: String, default: null },
+  email: { type: String, default: null },
+  phone: { type: String, default: null },
+  oauthID: { type: String, default: null, unique: true, sparse: true },
+  gender: [{ type: String, default: null }],
+  pronouns: [{ type: String, default: null }],
+  race: [{ type: String, default: null }],
```
Comment (alicesf2) [inline api/src/models/user.ts:62]: Change this to a string enum, you can view the options they currently have in the race section of the [Become a Contributor form](https://southsideweekly.com/contribute/)

**Review by alicesf2 — CHANGES_REQUESTED** [review]
A few User schema changes, fix those and you should be good to merge 💯

File: api/src/models/user.ts [hunk 4]
```diff
@@ -1,11 +1,109 @@
-import mongoose, { Document } from 'mongoose';
+import mongoose, { Document, Schema } from 'mongoose';
 
-interface User extends Document<any> {
+import { IPitch } from './pitch';
+
+/**
+ * Basic interface for a User Schema.
+ */
+interface IUser extends Document<any> {
```
Comment (alicesf2) [inline api/src/models/user.ts:8]: Is there a reason why this interface is different from the User schema below?

File: api/src/api/users.ts [hunk 5]
```diff
@@ -1,20 +1,135 @@
...
 
 const router = express.Router();
 
+/**
+ * Validates an ID to whether or not it is a valid MongoDB ID or not
+ * @param id Potential User ID to validate
+ */
+const isValidMongoId = (id: string): boolean => ObjectId.isValid(id);
+
+// gets all users
```
Comment (alicesf2) [inline api/src/api/users.ts:15]: Very very smol nit: Capitalize the first letter of the comment

File: api/src/api/users.ts [hunk 6]
```diff
@@ -1,20 +1,135 @@
...
     res.status(200).json({
-      message: `Successfully retrieved ${user.name} users.`,
+      message: `Successfully retrieved all users.`,
       success: true,
       result: user,
     });
   }),
 );
 
+//gets user by id
```
Comment (alicesf2) [inline api/src/api/users.ts:28]: Also add a space between the slashes and the comment (also capitalize)

File: api/src/api/users.ts [hunk 7]
```diff
@@ -1,20 +1,135 @@
...
+
+    res.status(200).json({
+      success: true,
+      message: 'Successfully updated user',
+      result: updatedUser,
+    });
+  }),
+);
+
+// deletes a user
```
Comment (alicesf2) [inline api/src/api/users.ts:105]: Also cap

**Review by alicesf2 — COMMENTED** [review]
Just one question and a few veryyy minor things, then you'll have my approval i promise

File: api/src/models/user.ts (hunk 4 above)
Comment (amit-sawhney) [reply api/src/models/user.ts:8]: I forgot to update it as I was changing the schema, my bad

**Review by alicesf2 — APPROVED** [review]
Just merge main and u should be good! 🚀

**Review by sahithimuthyala — APPROVED** [review]

**Review by alicesf2 — APPROVED** [review]
💯 🥇

**Pull request merged**